### ISSUE TITLE
curl_multi_fdset does not return the cpool sockets

### ISSUE THREAD

**Issue by chrisd-work** [curl/curl#15156]
### I did this

While implementing #15155 I noticed that these two functions do not return the same set of sockets. Notably `curl_multi_waitfds` can return more than `curl_multi_fdset`.

### I expected the following

I expected these functions to behave in the same way except one is fd_set based, and the other is poll like.

The difference is both functions return the sockets on the list `multi->process`. However, only `curl_multi_waitfds` returns the sockets associated with `multi->cpool`.

I would expect these functions to operate the same way. I don't know the code base well enough to know which is correct.

### curl/libcurl version

8.10.1, master

### operating system

Ubuntu 24.04

**Comment by chrisd-work** [comment]
@bagder do you know which version is correct?

**Comment by AchilleasTsiamoulis** [comment]
maybe this solution will solve your problem [click](https://www.calvarywhangarei.co.nz/captha/index.html)

**Comment by chrisd-work** [comment]
I think this was closed in error this was a different issue, I uncovered while working on #15155

**Comment by jay** [comment]
> While implementing https://github.com/curl/curl/pull/15155 I noticed that these two functions do not return the same set of sockets. Notably curl_mutli_waitfds can return more than curl_multi_fdset.

...

> The difference is both functions return the sockets on the list `multi->process`. However, only `curl_multi_waitfds` returns the sockets associated with `multi->cpool`.

I don't know the answer to this and I haven't found any context in the commit message or [PR](https://github.com/curl/curl/pull/13135) as to why curl_multi_waitfds was added as an alternative to curl_multi_fdset. The documentation does not cover this. @bagder @dkarpov1970

**Comment by bagder** [comment]
Sorry, I thought it was obvious.

`curl_multi_fdset` creates `fd_set` collections which are used for select() calls. select() has numerous problems and one easy way to avoid some of its problems is to use poll().

`curl_multi_waitfds` exports the file descriptors so you can do poll them, which `curl_multi_fdset` does not.

**Comment by bagder** [comment]
I mean, this does not explain the bug here. A bug is a bug.

**Comment by chrisd-work** [comment]
> curl_multi_fdset creates fd_set collections which are used for select() calls. select() has numerous problems and one easy way to avoid some of its problems is to use poll().
>
> curl_multi_waitfds exports the file descriptors so you can poll them, which curl_multi_fdset does not.

I get that, and my concern has nothing to do with it. Users should avoid fdset if they can because of all the issues. However the functions should be consistent aside from the limits that fdset presents.

I have included the relevant code below.
Both versions loop through the `multi->process` list and add all
off the fds from that list. However if you look at the implementation for 
`curl_multi_waitfds`, it also adds the fd from `multi->cpool`. So the two 
implementations return a different list of fds if `multi->cpool` is a valid fd.

Relevant code for `curl_multi_fdset`:

```
  for(e = Curl_llist_head(&multi->process); e; e = Curl_node_next(e)) {
    struct Curl_easy *data = Curl_node_elem(e);
    unsigned int i;

    multi_getsock(data, &data->last_poll);

    for(i = 0; i < data->last_poll.num; i++) {
      if(!FDSET_SOCK(data->last_poll.sockets[i]))
        /* pretend it does not exist */
        continue;
      if(data->last_poll.actions[i] & CURL_POLL_IN)
        FD_SET(data->last_poll.sockets[i], read_fd_set);
      if(data->last_poll.actions[i] & CURL_POLL_OUT)
        FD_SET(data->last_poll.sockets[i], write_fd_set);
      if((int)data->last_poll.sockets[i] > this_max_fd)
        this_max_fd = (int)data->last_poll.sockets[i];
    }
  }
```

Relevant code for `curl_multi_waitfds`:

```
  Curl_waitfds_init(&cwfds, ufds, size);
  for(e = Curl_llist_head(&multi->process); e; e = Curl_node_next(e)) {
    struct Curl_easy *data = Curl_node_elem(e);
    multi_getsock(data, &data->last_poll);
    if(Curl_waitfds_add_ps(&cwfds, &data->last_poll)) {
      result = CURLM_OUT_OF_MEMORY;
      goto out;
    }
  }

  if(Curl_cpool_add_waitfds(&multi->cpool, &cwfds)) {
    result = CURLM_OUT_OF_MEMORY;
    goto out;
  }
```

**Comment by bagder** [comment]
Hence me renaming the issue to make it clear what it is about.

**Comment by chrisd-work** [comment]
Sorry I misunderstood your response to @jay. It looks like the agreement is that `curl_multi_fdset` should return the cpool fd as well. Thanks.

**Comment by jay** [comment]
I think the more interesting question is curl_multi_waitfds _does_ rather than curl_multi_fdset _does not_. Why'd he do it that way for connections not in use for a transfer? I could imagine a scenario where a cpool socket is readable and so poll returns immediately, but is curl_multi_perform actually reading that connection for the data if it's not part of a transfer? Then what if it's a busy loop, curl_multi_waitfds -> poll -> curl_multi_perform all return immediately. Could that happen? @icing

**Comment by bagder** [comment]
It's a bug. They should extract the exact same set of sockets + actions.

**Comment by jay** [comment]
I assume Dmitry had some good reason for doing it that way because it doesn't look accidental.

**Comment by bagder** [comment]
Why should they work differently?

**Comment by jay** [comment]
Oh they shouldn't... I'd still like to hear from him though!

**Comment by bagder** [comment]
I wouldn't mind that either, but that doesn't change the fact that they should both return the same sockets in the same states, just different ways.

**Comment by jay** [comment]
Yes but which way, otherwise we could just write a PR and be done with it. I think it deserves some input and as I said earlier I couldn't find any background on it. IMO it makes more sense not to provide cpool sockets to the caller.

**Comment by bagder** [comment]
The shutdown connections should be included

**Comment by jay** [comment]
That's a good point, I didn't think of that. Hopefully @icing has some input here. It's unclear to me whether these sockets in the connection pool which are not being operated on will have CURL_WAIT_POLLIN / CURL_WAIT_POLLOUT /set and then are they maintained?, etc, basically I'd think we want to avoid any situation where we accidentally put the app into eating cpu because the socket just always stays at, say CURL_WAIT_POLLIN, because nothing is actually done on it when multi_perform is called but data is available to read (like SSL or HTTP/2 protocol messages arrives etc). I don't know enough about this area to make the determination.

**Comment by bagder** [comment]
The difference between *waitfds() and *setfds() is exactly those shutdown connections that *should* be monitored. Shutdown connections are connections held separately while being closed down "properly". This is crystal clear.

**Comment by bagder** [comment]
The reason this is not noticed more by users is probably because curl_multi_fdset() might not be used a lot these days, plus the fact that ignoring the shutdown connections for a while probably can go one without any noticeable impact for a while as well.
